File: redome-front/src/app/paciente/genotipo.comparado.dto.ts
import { BaseEntidade } from "../shared/base.entidade";
import { Locus } from './cadastro/exame/locus';
import { GenotipoDoadorComparadoDTO } from './genotipo.comparado.doador.dto';
import { GenotipoDTO } from './genotipo.dto';


/**
 * Classe que representa um genotipo, associada ao paciente.
 * @author Fillipe Queiroz
 */
export class GenotipoComparadoDTO extends BaseEntidade {

	private _rmr: number;
	private _nomePaciente: string;
	private _dataNascimento: Date;
	private _idade: string;
	private _sexo: string;
	private _peso: number;
	private _abo: string; 

	private _genotipoPaciente: GenotipoDTO[];
	private _genotiposDoadores: GenotipoDoadorComparadoDTO[]; 
	private _listaLocus: Locus[];

	/**
	 * @returns number
	 */
	public get rmr(): number {
		return this._rmr;
	}
	/**
	 * @param  {number} value
	 */
	public set rmr(value: number) {
		this._rmr = value;
	}
	/**
	 * @returns GenotipoDTO
	 */
	public get genotipoPaciente(): GenotipoDTO[] {
		return this._genotipoPaciente;
	}
	/**
	 * @param  {GenotipoDTO[]} value
	 */
	public set genotipoPaciente(value: GenotipoDTO[]) {
		this._genotipoPaciente = value;
	}

	public get genotiposDoadores(): GenotipoDoadorComparadoDTO[] {
		return this._genotiposDoadores;
	}

	public set genotiposDoadores(value: GenotipoDoadorComparadoDTO[]) {
		this._genotiposDoadores = value;
	}

    /**
     * Getter listaLocus
     * @return {Locus[]}
     */
	public get listaLocus(): Locus[] {
		return this._listaLocus;
	}

    /**
     * Setter listaLocus
     * @param {Locus[]} value
     */
	public set listaLocus(value: Locus[]) {
		this._listaLocus = value;
	}

	private pacientePossuiGenotipoComLocus(locus: Locus, comValor: boolean): boolean {
		if (this.genotipoPaciente) {
			return this.genotipoPaciente.some(genotipo => {
				let possui = genotipo.locus == locus.codigo;
				if (possui) {
					if (comValor) {
						if (!genotipo.primeiroAlelo || genotipo.primeiroAlelo == "") {
							possui = false;
						}
					}
					else {
						if (!genotipo.primeiroAlelo || genotipo.primeiroAlelo == "") {
							possui = true;
						}
						else {
							possui = false;
						}
					}					
				} 
				return possui;
			});			
		}
		return false; 
	}

	public pacientePossuiGenotipoComLocusEValor(locus: Locus): boolean {
		return this.pacientePossuiGenotipoComLocus(locus, true);
	}
	
	public pacientePossuiGenotipoComLocusESemValor(locus: Locus): boolean {
		return this.pacientePossuiGenotipoComLocus(locus, false);
	}


	public obterValorAleloGenotipoPaciente(codigoLocus: string): GenotipoDTO {
		if (this.genotipoPaciente) {
			return this.genotipoPaciente.filter(genotipo => genotipo.locus == codigoLocus)[0];
		}
	}

    /**
     * Getter dataNascimento
     * @return {Date}
     */
	public get dataNascimento(): Date {
		return this._dataNascimento;
	}

    /**
     * Setter dataNascimento
     * @param {Date} value
     */
	public set dataNascimento(value: Date) {
		this._dataNascimento = value;
	}

    /**
     * Getter sexo
     * @return {string}
     */
	public get sexo(): string {
		return this._sexo;
	}

    /**
     * Setter sexo
     * @param {string} value
     */
	public set sexo(value: string) {
		this._sexo = value;
	}

    /**
     * Getter peso
     * @return {number}
     */
	public get peso(): number {
		return this._peso;
	}

    /**
     * Setter peso
     * @param {number} value
     */
	public set peso(value: number) {
		this._peso = value;
	}

    /**
     * Getter abo
     * @return {string}
     */
	public get abo(): string {
		return this._abo;
	}

    /**
     * Setter abo
     * @param {string} value
     */
	public set abo(value: string) {
		this._abo = value;
	}

    /**
     * Getter idade
     * @return {string}
     */
	public get idade(): string {
		return this._idade;
	}

    /**
     * Setter idade
     * @param {string} value
     */
	public set idade(value: string) {
		this._idade = value;
	}

    /**
     * Getter nomePaciente
     * @return {string}
     */
	public get nomePaciente(): string {
		return this._nomePaciente;
	}

    /**
     * Setter nomePaciente
     * @param {string} value
     */
	public set nomePaciente(value: string) {
		this._nomePaciente = value;
	}
	
	

	public jsonToEntity(res: any) {
        throw new Error("Method not implemented.");
    }
}

File: redome/src/main/java/br/org/cancer/modred/controller/dto/GenotipoComparadoDTO.java
package br.org.cancer.modred.controller.dto;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

import br.org.cancer.modred.model.Locus;
import br.org.cancer.modred.redomelib.vo.ValorNmdpVo;

/**
 * Classe DTO para dados do genotipo entre paciente e doadores com match.
 * 
 * @author fillipe.queiroz
 *
 */
public class GenotipoComparadoDTO {

	private Long rmr;
	private String nomePaciente;
	private LocalDate dataNascimento;
	private String sexo;
	private BigDecimal peso;
	private String abo; 
	private String nomeMedicoResponsavel;
	private String nomeCentroAvaliador;
	private Integer idade;
	private List<ValorNmdpVo> valoresNmdp;
	
	private List<Locus> listaLocus;
	private List<GenotipoDTO> genotipoPaciente;
	private List<GenotipoDoadorComparadoDTO> genotiposDoadores;

	/**
	 * @return the rmr
	 */
	public Long getRmr() {
		return rmr;
	}

	/**
	 * @param rmr the rmr to set
	 */
	public void setRmr(Long rmr) {
		this.rmr = rmr;
	}

	/**
	 * @return the genotipoPaciente
	 */
	public List<GenotipoDTO> getGenotipoPaciente() {
		return genotipoPaciente;
	}

	/**
	 * @param genotipoPaciente the genotipoPaciente to set
	 */
	public void setGenotipoPaciente(List<GenotipoDTO> genotipoPaciente) {
		this.genotipoPaciente = genotipoPaciente;
	}

	/**
	 * @return the genotiposDoadores
	 */
	public List<GenotipoDoadorComparadoDTO> getGenotiposDoadores() {
		return genotiposDoadores;
	}

	/**
	 * @param genotiposDoadores the genotiposDoadores to set
	 */
	public void setGenotiposDoadores(List<GenotipoDoadorComparadoDTO> genotiposDoadores) {
		this.genotiposDoadores = genotiposDoadores;
	}

	
	/**
	 * @return the listaLocus
	 */
	public List<Locus> getListaLocus() {
		return listaLocus;
	}

	
	/**
	 * @param listaLocus the listaLocus to set
	 */
	public void setListaLocus(List<Locus> listaLocus) {
		this.listaLocus = listaLocus;
	}

	
	/**
	 * @return the dataNascimento
	 */
	public LocalDate getDataNascimento() {
		return dataNascimento;
	}

	
	/**
	 * @param dataNascimento the dataNascimento to set
	 */
	public void setDataNascimento(LocalDate dataNascimento) {
		this.dataNascimento = dataNascimento;
	}

	
	/**
	 * @return the sexo
	 */
	public String getSexo() {
		return sexo;
	}

	
	/**
	 * @param sexo the sexo to set
	 */
	public void setSexo(String sexo) {
		this.sexo = sexo;
	}

	
	/**
	 * @return the peso
	 */
	public BigDecimal getPeso() {
		return peso;
	}

	
	/**
	 * @param peso the peso to set
	 */
	public void setPeso(BigDecimal peso) {
		this.peso = peso;
	}

	
	/**
	 * @return the abo
	 */
	public String getAbo() {
		return abo;
	}

	
	/**
	 * @param abo the abo to set
	 */
	public void setAbo(String abo) {
		this.abo = abo;
	}

	
	/**
	 * @return the nomePaciente
	 */
	public String getNomePaciente() {
		return nomePaciente;
	}

	
	/**
	 * @param nomePaciente the nomePaciente to set
	 */
	public void setNomePaciente(String nomePaciente) {
		this.nomePaciente = nomePaciente;
	}

	/**
	 * @return the nomeMedicoResponsavel
	 */
	public String getNomeMedicoResponsavel() {
		return nomeMedicoResponsavel;
	}

	/**
	 * @param nomeMedicoResponsavel the nomeMedicoResponsavel to set
	 */
	public void setNomeMedicoResponsavel(String nomeMedicoResponsavel) {
		this.nomeMedicoResponsavel = nomeMedicoResponsavel;
	}

	/**
	 * @return the nomeCentroAvaliador
	 */
	public String getNomeCentroAvaliador() {
		return nomeCentroAvaliador;
	}

	/**
	 * @param nomeCentroAvaliador the nomeCentroAvaliador to set
	 */
	public void setNomeCentroAvaliador(String nomeCentroAvaliador) {
		this.nomeCentroAvaliador = nomeCentroAvaliador;
	}

	/**
	 * @return the idade
	 */
	public Integer getIdade() {
		return idade;
	}

	/**
	 * @param idade the idade to set
	 */
	public void setIdade(Integer idade) {
		this.idade = idade;
	}

	/**
	 * @return the valoresNmdp
	 */
	public List<ValorNmdpVo> getValoresNmdp() {
		return valoresNmdp;
	}

	/**
	 * @param valoresNmdp the valoresNmdp to set
	 */
	public void setValoresNmdp(List<ValorNmdpVo> valoresNmdp) {
		this.valoresNmdp = valoresNmdp;
	}
	
}
